neomodel/exceptions.py

```diff
@@ -53,15 +56,21 @@ def __init__(self, db_node, current_node_class_registry):
         """
         self.db_node = db_node
         self.current_node_class_registry = current_node_class_registry
-        
+
     def __str__(self):
         node_labels = ",".join(self.db_node.labels)
-        ncr_items = list(map(lambda x:"{} --> {}".format(",".join(x[0]), x[1]),
-                             self.current_node_class_registry.items()))
+        ncr_items = list(
```

Comment:
This is just a style/preference thing, but it might be more readable to do the following:
```
ncr_items_multiline = "\n".join(
    "{} --> {}".format(",".join(node_labels), model_class)
    for node_labels, model_class in self.current_node_class_registry.items()
)
```

Reply:
that definitely looks better, I use an autoformatter (black) that wreaked havoc on the existing formatting, hence the large amount of changes. 

Question is whether the implemented logic is solving your issues, can you check this?

Reply:
@mvanderkroon yes! I just came here to ask if it's ready for a test. I'll give it a go.

Reply:
I'm still getting a `NotALeaderError`. I can't see why, as I'm a bear of very little brain. I double-checked that the code I'm using is yours, and I can see the `except (ServiceUnavailableException, NotALeaderError, DatabaseUnavailableError` in your code, so that's very odd!

Here's the traceback:
```
2019-01-20 21:02:54.855 | INFO     | commands:do_second:80 - Attempting to write data to neo4j-2
2019-01-20 21:02:54.871 | ERROR    | commands:do_second:84 - Traceback (most recent call last):
  File "/commands.py", line 81, in do_second
    write_data()
  File "/usr/local/lib/python3.7/site-packages/neomodel/util.py", line 317, in wrapper
    return func(*args, **kwargs)
  File "/commands.py", line 56, in write_data
    return db.cypher_query(INSERT_QUERY)
  File "/usr/local/lib/python3.7/site-packages/neomodel/util.py", line 39, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/neomodel/util.py", line 261, in cypher_query
    raise exc_info[1].with_traceback(exc_info[2])
  File "/usr/local/lib/python3.7/site-packages/neomodel/util.py", line 233, in cypher_query
    results, meta = ([list(r.values()) for r in response], response.keys())
  File "/usr/local/lib/python3.7/site-packages/neomodel/util.py", line 233, in <listcomp>
    results, meta = ([list(r.values()) for r in response], response.keys())
  File "/usr/local/lib/python3.7/site-packages/neo4j/v1/api.py", line 736, in records
    self._session.fetch()
  File "/usr/local/lib/python3.7/site-packages/neo4j/v1/api.py", line 347, in fetch
    detail_count, _ = self._connection.fetch()
  File "/usr/local/lib/python3.7/site-packages/neo4j/bolt/connection.py", line 290, in fetch
    raise error
  File "/usr/local/lib/python3.7/site-packages/neo4j/bolt/connection.py", line 287, in fetch
    return self._fetch()
  File "/usr/local/lib/python3.7/site-packages/neo4j/bolt/connection.py", line 327, in _fetch
    response.on_failure(summary_metadata or {})
  File "/usr/local/lib/python3.7/site-packages/neo4j/v1/result.py", line 70, in on_failure
    raise CypherError.hydrate(**metadata)
neo4j.exceptions.NotALeaderError: No write operations are allowed directly on this database. Writes must pass through the leader. The role of this server is: FOLLOWER
```

Reply:
@robertlagrant, @mvanderkroon  There seem to be certain events that do not trigger a sync from the databsase server towards the driver. In the past, I have experienced this in relation to transactions where an operation was causing an error server-side, which however was only becoming known at client-side during a different transaction. This was confusing because it flagged errors at points that were not directly related to it.

Therefore, do you think that you could try [sync](https://neo4j.com/docs/api/python-driver/current/transactions.html#neo4j.neo4j.Session.Session.sync) on the session, at the point that is likely to be causing the problem? Perhaps this is one of those cases where the server has updated its state but this has not been propagated back to the client side (?). So, upon the first "fail", do a `sync` first and THEN retry the transaction. Hopefully this will update the client about which server has become the leader (?). Just an idea.

Reply:
@robertlagrant, @mvanderkroon Shall we include this change in an upcoming release? Did you guys try the `sync` before re-trying?

Reply:
@aanastasiou I am busy these weeks with other activities, so I was not able to test it just yet. At the earliest I'd be able to look into it at the beginning of next week.

Reply:
@mvanderkroon Of course, I understand, there is no rush, if I have sometime from my side I will give it a try too. 

All the best

Reply:
update from my end: we have been running into exactly the issue as @robertlagrant (countless `NOTALEADER` errors). However, in this case it seems unrelated to neomodel (or even the python-neo4j-driver): for us, causal clustering is just incredibly unstable. The causal cluster in fact gets stuck in a continuous LEADER election loop. I'm therefore quite certain that adding more `retry-logic` makes little sense on the application side. 

To me, 'causal clustering' is either fundamentally broken, or we are unlucky enough to encounter some sort of edge case bug. So far our experience with causal clustering has been far from ideal: unstable all the way through and write performance is ~6x worse than against unclustered. Our solution so far is to reduce the number of core nodes to 1 (essentially losing cluster capabilities) and keep our fingers crossed until we can migrate away from neo4j.

[fyi, seems related](https://github.com/neo4j/neo4j-python-driver/issues/276)

Reply:
@mvanderkroon Thanks for letting me know. @robertlagrant is there anything to add from your end on this (?).